### PR TITLE
Add Kafka message model

### LINKED ISSUE #119: Add database model for messages
The current session based implemention for messages is insufficient to support required functionality - most notably persistence on refresh.

- Create a BroadcastMessage model class that stores message text and timestamp.
- Update `kafka_consumer` admin command to populate messages into the BroadcastMessage table and to prune old messages from the list.
- On index page load, populate the Kafka message feed with all (or a default time window) of saved messages.

Not considered part of this issue but this implementation will also support auto-population of the message feed in the UI as it can support polling for messages newer than a given timestamp.

### PR DESCRIPTION
**Pull request by jamesturner246**
# Description

Add and use a new table model for persistent Kafka messages. I've named it `DruncMessage`, since `BroadcastMessage` is already taken by the druncschema package.

not sure if it's worth passing in timestamps separately to the view, so we can format them nicer. Something like [this](https://getbootstrap.com/docs/5.2/components/list-group/#custom-content).

Currently a temp `MESSAGE_EXPIRE_SECS = 60` to control this, so old messages begin purging after 60 seconds.

Pondering: is there ever a danger of `id` overflowing? How many messages are going to be chucked out here, over the whole period that the system is running?

Also the Django command only seems to run when in the 'process manager' page -- i.e. no stale deleting occurs unless we are on that page.

Fixes #119 

## Type of change

- [ ] Documentation (non-breaking change that adds or improves the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (non-breaking, back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (whatever its nature)

## Key checklist

- [x] All tests pass (eg. `python -m pytest`)
- [ ] The documentation builds and looks OK (eg. `python -m sphinx -b html docs docs/build`)
- [x] Pre-commit hooks run successfully (eg. `pre-commit run --all-files`)

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added or an issue has been opened to tackle that in the future. (Indicate issue here: # (issue))

### CODE REVIEW

**Review by cc-a — CHANGES_REQUESTED** [review]
Great. Just a few tweaks needed I think.

I'm not worried about the overflow if we ran the maths I think we'd probably find we have a few centuries to play with.

One thing that does give me pause is that under realistic scenarious there might be thousands of messages in the database so re-rendering all of them every second for multiple users might start to get noticeable on the server side. I think we should create an issue to look at some load testing and performance in deployment so we keep it in mind.

File: main/management/commands/kafka_consumer.py [hunk 1]
```diff
@@ -31,21 +32,46 @@ def handle(self, debug: bool = False, **kwargs: Any) -> None:  # type: ignore[mi
...
+                        # displayed by the web application
+                        DruncMessage.objects.bulk_create(
+                            [
+                                DruncMessage(timestamp=t, message=msg)
+                                for t, msg in zip(message_timestamps, message_bodies)
+                            ]
+                        )
+
+                # TODO: handle this (and timezone?) via settings.py
+                MESSAGE_EXPIRE_SECS = 60
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply main/management/commands/kafka_consumer.py:62]: Agree. [Some background](https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html), saying it's producer-side configured, so need to ask.

File: main/management/commands/kafka_consumer.py [hunk 2]
```diff
@@ -31,21 +32,46 @@ def handle(self, debug: bool = False, **kwargs: Any) -> None:  # type: ignore[mi
...
+                    # atomic here to prevent race condition with messages being
+                    # displayed by the web application
+                    query = DruncMessage.objects.filter(timestamp__lt=expire_time)
+                    if query.count():
+                        if debug:
+                            self.stdout.write(
+                                f"Deleting {query.count()} messages "
+                                f"older than {expire_time}."
+                            )
+                        query.delete()
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply main/management/commands/kafka_consumer.py:77]: Oops.. :+1:

File: process_manager/templates/process_manager/index.html
Comment (dalonsoa) [inline process_manager/templates/process_manager/index.html:1]: I thought this had been done in a different PR, so I'm not sure why it is here again.

File: process_manager/templates/process_manager/partials/message_items.html
Comment (dalonsoa) [inline process_manager/templates/process_manager/partials/message_items.html:1]: Same here.

File: process_manager/views/partials.py [hunk 3]
```diff
@@ -62,13 +61,8 @@ def process_table(request: HttpRequest) -> HttpResponse:
...
 def messages(request: HttpRequest) -> HttpResponse:
-    """Renders and pops Kafka messages from the user's session."""
-    with transaction.atomic():
-        # atomic to avoid race condition with kafka consumer
-        messages = request.session.load().get("messages", [])
-        request.session.pop("messages", [])
-        request.session.save()
-
+    """Renders Kafka messages from the database."""
+    messages = [f"{msg.timestamp}: {msg.message}" for msg in DruncMessage.objects.all()]
```
Comment (dalonsoa) [inline process_manager/views/partials.py:65]: In theory, we should be pulling the messages only relevant for this user, which means that there should be user-related information in the `DruncMessage` model, right?

**Review by dalonsoa — CHANGES_REQUESTED** [review]
It looks ok, but I think we have lost the user-specificity of the messages we had before. For account for that, the `DruncMessage` model will need to have an owner or a user or something like that, and add the messages from Kafka with that user information.

File: process_manager/views/partials.py (hunk 3 above)
Comment (cc-a) [reply process_manager/views/partials.py:65]: That was my initial assumption but after checking with the project team all users need to see all messages so there isn't a requirement for this.

File: process_manager/templates/process_manager/index.html
Comment (jamesturner246) [reply process_manager/templates/process_manager/index.html:1]: I remember it being done for the process table

File: process_manager/views/partials.py (hunk 3 above)
Comment (dalonsoa) [reply process_manager/views/partials.py:65]: Odd, but so be it.

**Comment by jamesturner246** [comment]
Done. Managed to do the timezone stuff with django-only fns. @dalonsoa, are your comments addressed?

**Review by dalonsoa — APPROVED** [review]
QA checks are failing, but otherwise the code looks good!

File: main/management/commands/kafka_consumer.py [hunk 4]
```diff
@@ -31,21 +31,37 @@ def handle(self, debug: bool = False, **kwargs: Any) -> None:  # type: ignore[mi
...
             for messages in consumer.poll(timeout_ms=500).values():
+                message_timestamps = []
                 message_bodies = []
                 for message in messages:
                     if debug:
                         self.stdout.write(f"Message received: {message}")
                         self.stdout.flush()
+
+                    # Convert Kafka timestamp (milliseconds) to datetime (seconds).
+                    timestamp = datetime.fromtimestamp(message.timestamp / 1e3, tz=UTC)
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply main/management/commands/kafka_consumer.py:42]: :thinking: 

```
<ipython-input-11-d286e0fff8ad>:1: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
```

Trend is towards overdefinedness.

[the remark this answers is not in the document]
Comment (cc-a) [reply main/management/commands/kafka_consumer.py:42]: Ah, fair enough.

File: process_manager/views/partials.py [hunk 5]
```diff
@@ -62,12 +62,11 @@ def process_table(request: HttpRequest) -> HttpResponse:
...
-    """Renders and pops Kafka messages from the user's session."""
-    with transaction.atomic():
-        # atomic to avoid race condition with kafka consumer
-        messages = request.session.load().get("messages", [])
-        request.session.pop("messages", [])
-        request.session.save()
+    """Renders Kafka messages from the database."""
+    messages = []
+    for msg in DruncMessage.objects.all():
+        timestamp = localtime(msg.timestamp).strftime("%Y-%m-%d %H:%M:%S")
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply process_manager/views/partials.py:69]: :+1:  Done.

**Comment by jamesturner246** [comment]
Sorry, should have asked first: do you prefer to do the merging in future @cc-a ?
